python-api/gstswitch/server.py

```diff
@@ -22,6 +24,107 @@
 TOOLS_DIR = '/'.join(os.getcwd().split('/')[:-1]) + '/tools/'
 
 
+class ProcessMonitor(object):
+    """Runs a Command in a Background-Thread and monitors it's output
+
+    Can block until the Command prints a certain string and log the full
+    output into a file
+    """
+
+    def __init__(self, cmd, logtarget=sys.stderr):
+        self.log = logging.getLogger('server-output-monitor')
+
+        # After calling start(), _proc contains a subprocess.Popen instance
+        self._proc = None
+
+        # Command to run (Array passed to subprocess.Popen)
+        self._cmd = cmd
+
+        # Logfile to write to
+        self._logtarget = logtarget
+
+        # Internal Buffer to search for matched when wait_for_output is called
+        self._buffer = ""
+
+    def start(self):
+        """Starts up the process specified in the constructor in a thread.
+        Blocks until the process is started and has a pid assigned, but not
+        longer then 0.5s.
+        """
+
+        assert self._proc is None
+        self.log.debug("starting subprocess")
+
+        self._proc = subprocess.Popen(
+            self._cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            bufsize=0,
+            shell=False)
```

Comment:
The writing to a PIPE and then using select below has a few problem, the main one being that the whole server can lock up if it tries to write to stdout and the pipe's buffer is full. Another problem here is the output getting buffered and not available until a large enough amount is written.

The work around for the partial buffering problem is to use ptys like follows;
```python
import pty
master, slave = pty.openpty()

p = Popen(cmd, shell=True, stdin=PIPE, stdout=slave, stderr=slave, close_fds=True)
stdout = os.fdopen(master)
```

An alternative solution is to get the subprocess to write to a file. You'll want to open the file in the required buffered mode (probably line) and then pass it to subprocess. Then open the file in another file object and read it at your leisure. You can use tempfile.NamedTemporaryFile to create it.